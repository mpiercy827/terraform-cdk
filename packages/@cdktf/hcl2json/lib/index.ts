// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference lib="dom" />

// Inspired by
// https://github.com/ts-terraform/ts-terraform
// https://github.com/aaronpowell/webpack-golang-wasm-async-loader

import fs from "fs-extra";
import path from "path";
import { Go } from "./wasm_exec";
import { deepMerge } from "./deepmerge";
import { gunzipSync } from "zlib";

interface GoBridge {
  parse: (filename: string, hcl: string) => Promise<string>;
  getReferencesInExpression: (filename: string, hcl: string) => Promise<string>;
}

// eslint-disable-next-line @typescript-eslint/ban-types
const jsRoot: Record<string, Function> = {};

function sleep() {
  return new Promise((resolve) => {
    setTimeout(resolve, 0);
  });
}

function goBridge(getBytes: Promise<Buffer>) {
  let ready = false;

  async function init() {
    const go = new Go();
    const bytes = await getBytes;
    const result = await WebAssembly.instantiate(bytes, go.importObject);
    void go.run(result.instance, { __parse_terraform_config_wasm__: jsRoot });
    ready = true;
  }

  init().catch((error) => {
    throw error;
  });

  const proxy = new Proxy({} as GoBridge, {
    get: (_, key: string) => {
      return async (...args: unknown[]) => {
        while (!ready) {
          await sleep();
        }

        if (!(key in jsRoot)) {
          throw new Error(
            `There is nothing defined with the name "${key.toString()}"`
          );
        }

        if (typeof jsRoot[key] !== "function") {
          return jsRoot[key];
        }

        return new Promise((resolve, reject) => {
          const cb = (err: string, ...msg: string[]) =>
            // @ts-ignore
            err ? reject(new Error(err)) : resolve(...msg);

          const run = () => {
            jsRoot[key].apply(undefined, [...args, cb]);
          };

          run();
        });
      };
    },
  });

  return proxy;
}

const loadWasm = async () => {
  return gunzipSync(
    await fs.readFile(path.join(__dirname, "..", "main.wasm.gz"))
  );
};

const wasm = goBridge(loadWasm());

export async function parse(
  filename: string,
  contents: string
): Promise<Record<string, any>> {
  const res = await wasm.parse(filename, contents);
  return JSON.parse(res);
}

export async function convertFiles(
  workingDirectory: string
): Promise<Record<string, any> | void> {
  let tfFileContents = "";
  const tfJSONFileContents: Record<string, any>[] = [];

  for (const file of fs.readdirSync(workingDirectory)) {
    const filePath = path.resolve(workingDirectory, file);
    if (!fs.lstatSync(filePath).isDirectory()) {
      if (file.match(/\.tf$/)) {
        tfFileContents += fs.readFileSync(filePath, "utf-8");
        tfFileContents += "\n";
      } else if (file.match(/\.tf\.json$/)) {
        tfJSONFileContents.push(JSON.parse(fs.readFileSync(filePath, "utf-8")));
      }
    }
  }

  if (tfFileContents === "" && tfJSONFileContents === []) {
    console.error(`No '.tf' or '.tf.json' files found in ${workingDirectory}`);
    return;
  }

  return deepMerge(
    await parse("hcl2json.tf", tfFileContents),
    ...tfJSONFileContents
  );
}

type CodeMarker = {
  Byte: number;
  Line: number;
  Column: number;
};
type Range = {
  End: CodeMarker;
  Start: CodeMarker;
};
type TerraformTraversalPart = {
  Name: string;
  SrcRange: Range;
};
// Reference to a variable / module / resource
type TerraformTraversal = { Traversal: TerraformTraversalPart[] };
type TerraformFunctionCall = {
  Args: TerraformObject[];
  Name: string;
  ExpandFinal: boolean;
  NameRange: Range;
  OpenParenRange: Range;
  CloseParenRange: Range;
};
type TerraformLiteral = {
  SrcRange: Range;
  Val: unknown; // No value is passed down, we ignore them
};
type TerraformEmbeddedExpression = {
  Wrapped: TerraformObject;
};
type TerraformExpression = {
  Parts: TerraformObject[];
};
type TerraformObject =
  | TerraformEmbeddedExpression
  | TerraformExpression
  | TerraformFunctionCall
  | TerraformLiteral
  | TerraformTraversal;

type GoExpressionParseResult = null | TerraformObject;

type Reference = { value: string; startPosition: number; endPosition: number };

function traversalToReference(traversal: TerraformTraversal): Reference {
  if (
    traversal.Traversal.find(
      (part) => part.SrcRange.Start.Line !== 0 || part.SrcRange.End.Line !== 0
    )
  ) {
    throw new Error(`Parsed references are expected to be a single line`);
  }

  return {
    value: traversal.Traversal.map((part) => part.Name).join("."),
    startPosition: traversal.Traversal[0].SrcRange.Start.Column,
    endPosition:
      traversal.Traversal[traversal.Traversal.length - 1].SrcRange.End.Column,
  };
}

function findAllReferencesInAst(entry: TerraformObject): Reference[] {
  if ("Traversal" in entry) {
    return [traversalToReference(entry)];
  }

  if ("Parts" in entry) {
    return entry.Parts.reduce(
      (carry, part) => [...carry, ...findAllReferencesInAst(part)],
      [] as Reference[]
    );
  }

  if ("Wrapped" in entry) {
    return findAllReferencesInAst(entry.Wrapped);
  }

  if ("Args" in entry) {
    return entry.Args.reduce(
      (carry, arg) => [...carry, ...findAllReferencesInAst(arg)],
      [] as Reference[]
    );
  }

  return [];
}

export async function getReferencesInExpression(
  filename: string,
  expression: string
): Promise<Reference[]> {
  const res = await wasm.getReferencesInExpression(
    filename,
    JSON.stringify(expression)
  );
  const ast = JSON.parse(res) as GoExpressionParseResult;

  if (!ast) {
    return [];
  }

  return findAllReferencesInAst(ast);
}
