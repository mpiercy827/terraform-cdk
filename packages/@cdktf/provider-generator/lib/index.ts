export * from "./get/constructs-maker";
export * as config from "./config";
export { getResourceNamespace } from "./get/generator/constants/provider-namespaces";
export { isRegistryModule } from "./get/module";
export { TerraformModuleConstraint, isLocalModule } from "./config";
export {
  readSchema,
  ProviderSchema,
  Schema,
  BlockType,
  AttributeType,
  Block,
} from "./get/generator/provider-schema";

import { CodeMaker } from "codemaker";
import * as srcmak from "jsii-srcmak";
import { generateJsiiLanguage } from "./get/constructs-maker";
import { TerraformProviderGenerator } from "./get/generator/provider-generator";
export { setLogger } from "./config";

// Used for testing only
export async function generateProviderBindingsFromSchema(
  targetPath: string,
  schemaJSON: any,
  options?: srcmak.Options
): Promise<void> {
  const code = new CodeMaker();
  new TerraformProviderGenerator(code, schemaJSON);
  await code.save(targetPath);

  if (options) {
    await generateJsiiLanguage(code, options);
  }
}
