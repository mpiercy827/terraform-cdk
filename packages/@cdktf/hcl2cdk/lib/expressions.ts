import * as t from "@babel/types";
import reservedWords from "reserved-words";
import { camelCase, pascalCase } from "./utils";
import { TerraformResourceBlock, Scope } from "./types";
import { getResourceNamespace } from "@cdktf/provider-generator";
import { getReferencesInExpression } from "@cdktf/hcl2json";

export type Reference = {
  start: number;
  end: number;
  referencee: { id: string; full: string }; // identifier for resource
  useFqn?: boolean;
  isVariable?: boolean;
};

const DOLLAR_REGEX = /\$/g;

export async function extractReferencesFromExpression(
  input: string,
  nodeIds: readonly string[],
  scopedIds: readonly string[] = [] // dynamics introduce new scoped variables that are not the globally accessible ids
): Promise<Reference[]> {
  const possibleVariableSpots = await getReferencesInExpression(
    "main.tf",
    input
  );

  return possibleVariableSpots.reduce((carry, spot) => {
    const { value, startPosition, endPosition } = spot;
    // no reference
    if (
      !value.includes(".") || // just a literal
      value.startsWith(".") || // dangling property access
      value.endsWith("...") || // spread (likely in for loop)
      value.startsWith("count.") || // count variable
      value.startsWith("each.") || // each variable
      // https://www.terraform.io/docs/language/expressions/references.html#filesystem-and-workspace-info
      value.startsWith("path.module") ||
      value.startsWith("path.root") ||
      value.startsWith("path.cwd") ||
      value.startsWith("terraform.workspace")
    ) {
      return carry;
    }

    const referenceParts = value.split(".");

    const corespondingNodeId = [...nodeIds, ...scopedIds].find((id) => {
      const parts = id.split(".");
      const matchesFirst = parts[0] === referenceParts[0];
      const matchesFirstTwo =
        matchesFirst && (parts[1] === referenceParts[1] || parts.length === 1);

      return (
        matchesFirstTwo &&
        (parts[0] === "data" ? parts[2] === referenceParts[2] : true)
      );
    });

    if (!corespondingNodeId) {
      // This is most likely a false positive, so we just ignore it
      // We include the log below to help debugging
      console.error(
        `Found a reference that is unknown: ${input} has reference "${value}". The id was not found in ${JSON.stringify(
          nodeIds
        )} with temporary values ${JSON.stringify(scopedIds)}.
        Please leave a comment at https://cdk.tf/bugs/convert-expressions if you run into this issue.`
      );
      return carry;
    }

    if (scopedIds.includes(corespondingNodeId)) {
      return carry;
    }

    const spotParts = value.split(".");
    let isThereANumericAccessor = false;
    const referenceSpotParts = spotParts.filter((part) => {
      if (!Number.isNaN(parseInt(part, 10))) {
        isThereANumericAccessor = true;
        return false;
      }

      return !isThereANumericAccessor;
    });
    const fullReference = isThereANumericAccessor
      ? referenceSpotParts.slice(0, 2).join(".")
      : value;

    const isVariable = value.startsWith("var.");
    const useFqn =
      // Can not use FQN on vars
      !isVariable &&
      // Can not use FQN on locals
      !value.startsWith("local.") &&
      // If the following character is
      (input.substr(endPosition + 1, 1) === "*" || // a * (splat) we need to use the FQN
        input.substr(endPosition, 1) === "[" || // a property access
        isThereANumericAccessor || // a numeric access
        fullReference.split(".").length < 3);

    const ref: Reference = {
      start: startPosition,
      end: endPosition,
      referencee: {
        id: corespondingNodeId,
        full: fullReference,
      },
      useFqn,
      isVariable,
    };
    return [...carry, ref];
  }, [] as Reference[]);
}

export function referenceToVariableName(scope: Scope, ref: Reference): string {
  const parts = ref.referencee.full.split(".");
  const resource = parts[0] === "data" ? `${parts[0]}.${parts[1]}` : parts[0];
  const name = parts[0] === "data" ? parts[2] : parts[1];
  return variableName(scope, resource, name);
}

function validVarName(name: string) {
  if (reservedWords.check(name)) {
    return `${name}Var`;
  }

  if (!Number.isNaN(parseInt(name[0], 10))) {
    return `d${name}`;
  }

  return name;
}

export function variableName(
  scope: Scope,
  resource: string,
  name: string
): string {
  // name collision, we need to prefix the name
  if (scope.variables[name]) {
    if (resource === scope.variables[name].resource) {
      return scope.variables[name].variableName;
    }

    // we only cache one per name
    return validVarName(camelCase([resource, name].join("_")));
  }

  const variableName = validVarName(
    camelCase(
      ["var", "local", "module"].includes(resource)
        ? name
        : [resource, name].join("_")
    )
  );

  scope.variables[name] = { variableName, resource };
  return variableName;
}

export function constructAst(type: string, isModuleImport: boolean) {
  if (isModuleImport) {
    return t.memberExpression(t.identifier(type), t.identifier(type));
  }

  // resources or data sources
  if (!type.includes("./") && type.includes(".")) {
    const parts = type.split(".");
    if (parts[0] === "data") {
      const [, provider, resource] = parts;
      const namespace = getResourceNamespace(provider, resource);
      if (namespace) {
        return t.memberExpression(
          t.memberExpression(
            t.identifier(provider), // e.g. aws
            t.identifier(namespace.name) // e.g. EC2
          ),
          t.identifier(pascalCase(`data_${provider}_${resource}`)) // e.g. DataAwsInstance
        );
      }

      return t.memberExpression(
        t.identifier(provider), // e.g. aws
        t.identifier(pascalCase(`data_${provider}_${resource}`)) // e.g. DataAwsNatGateway
      );
    }

    const [provider, resource] = parts;
    const namespace = getResourceNamespace(provider, resource);
    if (namespace) {
      return t.memberExpression(
        t.memberExpression(
          t.identifier(provider), // e.g. aws
          t.identifier(namespace.name) // e.g. EC2
        ),
        t.identifier(pascalCase(resource)) // e.g. Instance
      );
    }
    return t.memberExpression(
      t.identifier(provider), // e.g. google
      t.identifier(pascalCase(resource)) // e.g. BigQueryTable
    );
  }

  return t.identifier(pascalCase(type));
}

export function referenceToAst(scope: Scope, ref: Reference) {
  const [resource, , ...selector] = ref.referencee.full.split(".");

  const variableReference = t.identifier(
    camelCase(referenceToVariableName(scope, ref))
  );

  if (resource === "data") {
    selector.shift(); // remove the data part so that the name is not used in the selector
  }

  const accessor = selector.reduce(
    (carry, member, index) =>
      t.memberExpression(
        carry,
        t.identifier(
          index === 0 && resource === "module"
            ? camelCase(member + "Output")
            : camelCase(member)
        )
      ),
    variableReference as t.Expression
  );

  if (ref.useFqn) {
    return t.memberExpression(accessor, t.identifier("fqn"));
  }

  if (ref.isVariable) {
    return t.memberExpression(accessor, t.identifier("value"));
  }
  return accessor;
}

export function referencesToAst(
  scope: Scope,
  input: string,
  refs: Reference[],
  scopedIds: readonly string[] = [] // dynamics introduce new scoped variables that are not the globally accessible ids
): t.Expression {
  if (refs.length === 0) {
    return t.stringLiteral(input);
  }

  const refAsts = refs
    .sort((a, b) => a.start - b.start)
    .filter((ref) => !scopedIds.includes(ref.referencee.id))
    .map((ref) => ({ ref, ast: referenceToAst(scope, ref) }));

  if (
    refAsts.length === 1 &&
    refAsts[0].ref.start === "${".length &&
    refAsts[0].ref.end === input.length - "}".length &&
    !refAsts[0].ref.useFqn
  ) {
    return refAsts[0].ast;
  }

  // string parts in the template string
  const quasis: t.TemplateElement[] = [];
  // dynamic values in the template string
  const expressions: t.Expression[] = [];

  let lastEnd = 0;

  refAsts.forEach(({ ref, ast }) => {
    // leading quasi
    if (ref.start !== lastEnd) {
      quasis.push(
        t.templateElement({
          raw: input.substring(lastEnd, ref.start).replace(DOLLAR_REGEX, "\\$"),
        })
      );
    }

    expressions.push(ast);

    lastEnd = ref.end;
  });

  // trailing quasi
  quasis.push(
    t.templateElement(
      {
        raw: input
          .substring(lastEnd, input.length)
          .replace(DOLLAR_REGEX, "\\$"),
      },
      true
    )
  );

  return t.templateLiteral(quasis, expressions);
}

export type DynamicBlock = {
  path: string;
  for_each: string;
  content: TerraformResourceBlock;
  scopedVar: string;
};
export const extractDynamicBlocks = (
  config: TerraformResourceBlock,
  path = ""
): DynamicBlock[] => {
  if (typeof config !== "object") {
    return [];
  }

  if (!config) {
    return [];
  }

  if (Array.isArray(config)) {
    return config.reduce(
      (carry, item, index) => [
        ...carry,
        ...extractDynamicBlocks(item, `${path}.${index}`),
      ],
      []
    );
  }

  if ("dynamic" in config) {
    const dynamic = (config as any).dynamic;
    const scopedVar = Object.keys(dynamic)[0];
    const { for_each, content } = dynamic[scopedVar][0];

    return [
      {
        path: `${path}.${scopedVar}`,
        for_each,
        content,
        scopedVar,
      },
    ];
  }

  return Object.entries(config).reduce((carry, [key, value]) => {
    return [...carry, ...extractDynamicBlocks(value as any, `${path}.${key}`)];
  }, [] as DynamicBlock[]);
};

export async function findUsedReferences(
  nodeIds: string[],
  item: TerraformResourceBlock
): Promise<Reference[]> {
  if (typeof item === "string") {
    return await extractReferencesFromExpression(item, nodeIds, []);
  }

  if (typeof item !== "object" || item === null || item === undefined) {
    return [];
  }

  if (Array.isArray(item)) {
    return (
      await Promise.all(item.map((i) => findUsedReferences(nodeIds, i), []))
    ).reduce((carry, i) => [...carry, ...i], []);
  }

  if (item && "dynamic" in item) {
    const dyn = (item as any)["dynamic"];
    const { for_each, ...others } = dyn;
    const dynamicRef = Object.keys(others)[0];
    return await findUsedReferences([...nodeIds, dynamicRef], dyn);
  }

  return (
    await Promise.all(
      Object.values(item as Record<string, any>).map(
        (i) => findUsedReferences(nodeIds, i),
        []
      )
    )
  ).reduce((carry, i) => [...carry, ...i], []);
}

// This only guesses if the type of an expression is list, it should be replaced by something that understands
// the type of the expression, solved by https://github.com/hashicorp/terraform-cdk/issues/842
export function isListExpression(item: string) {
  const hasListExtension =
    item.includes("[") &&
    item.includes("for ") &&
    item.includes(" in ") &&
    item.includes("]");

  if (!hasListExtension) {
    return false;
  }

  // We might have wrapped it in a function that collapses the list
  return !["element", "index", "length", "lookup", "one", "join"].some(
    (collapsingTfFunction) =>
      item.includes(`${collapsingTfFunction}(`) &&
      item.indexOf(`${collapsingTfFunction}(`) < item.indexOf("for")
  );
}
