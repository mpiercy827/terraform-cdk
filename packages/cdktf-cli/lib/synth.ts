import { SynthesizedStack } from "./synth-stack";
import { AnnotationMetadataEntryType, CONTEXT_ENV } from "cdktf";
import * as chalk from "chalk";
import { DEPRECATED_FLAGS } from "cdktf/lib/features";

const chalkColour = new chalk.Instance();

/**
 * prints annotations via console
 * @throws Error if an Annotation with level === ERROR has been encountered
 */
export function printAnnotations(stacks: SynthesizedStack[]) {
  let encounteredAnnotationError = false;

  stacks.forEach((stack) =>
    (stack.annotations || []).forEach((annotation) => {
      if (annotation.level === AnnotationMetadataEntryType.ERROR)
        encounteredAnnotationError = true;

      let color = chalkColour.whiteBright;
      let level: string = annotation.level;
      switch (annotation.level) {
        case AnnotationMetadataEntryType.INFO: {
          level = "INFO";
          break;
        }
        case AnnotationMetadataEntryType.WARN: {
          color = chalkColour.hex("#FFA500"); // orange
          level = "WARN";
          break;
        }
        case AnnotationMetadataEntryType.ERROR: {
          color = chalkColour.redBright;
          level = "ERROR";
          break;
        }
      }
      console.log(
        color(`${level} [${annotation.constructPath}]: ${annotation.message}`)
      );
      if (
        annotation.level === AnnotationMetadataEntryType.ERROR &&
        annotation.stacktrace
      )
        console.log(chalkColour.gray(annotation.stacktrace.join("\n")));
    })
  );

  if (encounteredAnnotationError) {
    throw new Error(
      "While synthesizing one or more error annotations have been encountered. Please check the log output above."
    );
  }
}

export function printWarningsForDeprecatedFlags() {
  const context: Record<string, boolean | undefined> = JSON.parse(
    process.env[CONTEXT_ENV] || "{}"
  );

  // We want to print a warning for every deprecated flag that is set to false or not set
  const deprecatedFlagsInUse = DEPRECATED_FLAGS.filter(
    (flag) => context[flag] === false || context[flag] === undefined
  );

  if (deprecatedFlagsInUse.length > 0) {
    const plural = deprecatedFlagsInUse.length > 1;
    console.log(
      chalkColour.yellow(
        `WARNING: The flag${plural ? "s" : ""} ${deprecatedFlagsInUse.join(
          " & "
        )} ${
          plural ? "are" : "is"
        } deprecated and will be removed in the next minor release. Please set ${
          plural ? "them" : "it"
        } to true in your cdktf.json's context map.`
      )
    );
  }
}
