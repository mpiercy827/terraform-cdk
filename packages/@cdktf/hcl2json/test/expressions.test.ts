import { getReferencesInExpression } from "../lib";

function parseExpression(expr: string) {
  return () =>
    expect(
      getReferencesInExpression("main.tf", expr)
    ).resolves.toMatchSnapshot();
}

describe("getReferencesInExpression", () => {
  test("parses a simple string", parseExpression("foo"));
  test("parses a reference string", parseExpression("${var.foo}"));
  test("parses three part references", parseExpression("${module.foo.output}"));
  test(
    "parses four part references",
    parseExpression("${data.type.name.attr}")
  );
  test(
    "parses a terraform function",
    parseExpression('${replace(module.foo.output, "-", var.bar)}')
  );

  test(
    "parses nested terraform functions",
    parseExpression(
      '${split(var.separator, lower(replace(module.foo.output, "-", var.bar)))}'
    )
  );

  test(
    "parses embedded vars",
    parseExpression("Hey, did you hear about ${module.foo.output}s new album?")
  );
  test(
    "parses lists",
    parseExpression('[5, ${module.foo.output}, "val", true]')
  );

  test(
    "parses maps",
    parseExpression(
      '{ x = ${module.foo.output}, y = "val", z = true, a = ${replace(module.foo.output, "-", var.bar)} }'
    )
  );

  test("fails on malformed expressions", () => {
    expect(
      getReferencesInExpression("main.tf", "${module.foo.output")
    ).rejects.toMatchSnapshot();
  });
});
