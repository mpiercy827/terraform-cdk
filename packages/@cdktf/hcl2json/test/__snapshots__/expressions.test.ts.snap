// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReferencesInExpression fails on malformed expressions 1`] = `[Error: :0,0-2: Unclosed template interpolation sequence; There is no closing brace for this interpolation sequence before the end of the file. This might be caused by incorrect nesting inside the given expression.]`;

exports[`getReferencesInExpression parses a reference string 1`] = `
Array [
  Object {
    "endPosition": 9,
    "startPosition": 2,
    "value": "var.foo",
  },
]
`;

exports[`getReferencesInExpression parses a simple string 1`] = `Array []`;

exports[`getReferencesInExpression parses a terraform function 1`] = `
Array [
  Object {
    "endPosition": 27,
    "startPosition": 10,
    "value": "module.foo.output",
  },
  Object {
    "endPosition": 41,
    "startPosition": 34,
    "value": "var.bar",
  },
]
`;

exports[`getReferencesInExpression parses embedded vars 1`] = `
Array [
  Object {
    "endPosition": 43,
    "startPosition": 26,
    "value": "module.foo.output",
  },
]
`;

exports[`getReferencesInExpression parses four part references 1`] = `
Array [
  Object {
    "endPosition": 21,
    "startPosition": 2,
    "value": "data.type.name.attr",
  },
]
`;

exports[`getReferencesInExpression parses lists 1`] = `
Array [
  Object {
    "endPosition": 23,
    "startPosition": 6,
    "value": "module.foo.output",
  },
]
`;

exports[`getReferencesInExpression parses maps 1`] = `
Array [
  Object {
    "endPosition": 25,
    "startPosition": 8,
    "value": "module.foo.output",
  },
  Object {
    "endPosition": 80,
    "startPosition": 63,
    "value": "module.foo.output",
  },
  Object {
    "endPosition": 94,
    "startPosition": 87,
    "value": "var.bar",
  },
]
`;

exports[`getReferencesInExpression parses nested terraform functions 1`] = `
Array [
  Object {
    "endPosition": 21,
    "startPosition": 8,
    "value": "var.separator",
  },
  Object {
    "endPosition": 54,
    "startPosition": 37,
    "value": "module.foo.output",
  },
  Object {
    "endPosition": 68,
    "startPosition": 61,
    "value": "var.bar",
  },
]
`;

exports[`getReferencesInExpression parses three part references 1`] = `
Array [
  Object {
    "endPosition": 19,
    "startPosition": 2,
    "value": "module.foo.output",
  },
]
`;
